compiling ROMIO with PVFS2 support

compiling the ompi master and enabling PVFS2 support in ROMIO will lead to an error at compile time

adio/ad_pvfs2/ad_pvfs2_aio.c: In function 'ADIOI_PVFS2_AIO_contig':

adio/ad_pvfs2/ad_pvfs2_aio.c:126:6: error: implicit declaration of function 'PMPIX_Grequest_class_create' [-Werror=implicit-function-declaration]
      MPIX_Grequest_class_create(ADIOI_GEN_aio_query_fn,
      ^

adio/ad_pvfs2/ad_pvfs2_aio.c:131:2: error: implicit declaration of function 'PMPIX_Grequest_class_allocate' [-Werror=implicit-function-declaration]
  MPIX_Grequest_class_allocate(ADIOI_PVFS2_greq_class, aio_req, request);
  ^


Note, that Open MPI sets the --disable-aio flag when configuring ROMIO. My suspicion is that there is 
some protection through macros (HAVE_MPI_GREQUEST_EXTENSIONS or one of the AIO macros) is missing from this file, but not entirely sure.

@roblatham00 Any ideas/suggestions how to resolve that?

I thought we fixed this years ago?  Ah, wait, we set the function pointers
correctly last year, but we are probably still building the (unused) aio
routines.   I'll generate a pull request.

==rob

On Wed, Feb 25, 2015 at 9:39 AM, Edgar Gabriel <notifications@github.com>
wrote:

> compiling the ompi master and enabling PVFS2 support in ROMIO will lead to
> an error at compile time
>
> adio/ad_pvfs2/ad_pvfs2_aio.c: In function 'ADIOI_PVFS2_AIO_contig':
>
> adio/ad_pvfs2/ad_pvfs2_aio.c:126:6: error: implicit declaration of
> function 'PMPIX_Grequest_class_create'
> [-Werror=implicit-function-declaration]
> MPIX_Grequest_class_create(ADIOI_GEN_aio_query_fn,
> ^
>
> adio/ad_pvfs2/ad_pvfs2_aio.c:131:2: error: implicit declaration of
> function 'PMPIX_Grequest_class_allocate'
> [-Werror=implicit-function-declaration]
> MPIX_Grequest_class_allocate(ADIOI_PVFS2_greq_class, aio_req, request);
> ^
>
> Note, that Open MPI sets the --disable-aio flag when configuring ROMIO. My
> suspicion is that there is
> some protection through macros (HAVE_MPI_GREQUEST_EXTENSIONS or one of the
> AIO macros) is missing from this file, but not entirely sure.
>
> @roblatham00 <https://github.com/roblatham00> Any ideas/suggestions how
> to resolve that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/open-mpi/ompi/issues/426>.
>

--- a/ompi/mca/io/romio/romio/adio/ad_pvfs2/ad_pvfs2_aio.c
+++ b/ompi/mca/io/romio/romio/adio/ad_pvfs2/ad_pvfs2_aio.c
@@ -121,6 +121,7 @@ void ADIOI_PVFS2_AIO_contig(ADIO_File fd, void *buf, int count,
     }
     /* --END ERROR HANDLING-- */
 
+#ifdef HAVE_MPI_GREQUEST_EXTENSIONS
     /* posted. defered completion */
     if (ret == 0) { 
 	if (ADIOI_PVFS2_greq_class == 0) {
@@ -132,6 +133,15 @@ void ADIOI_PVFS2_AIO_contig(ADIO_File fd, void *buf, int count,
 	MPIX_Grequest_class_allocate(ADIOI_PVFS2_greq_class, aio_req, request);
 	memcpy(&(aio_req->req), request, sizeof(*request));
     }
+#else
+    /* if generalized request extensions not available, we will have to process
+     * this operation right here */
+    int error;
+    ret = PVFS_sys_wait(aio_req->op_id, "ADIOI_PVFS2_AIO_Contig", &error);
+    if (ret == 0) {
+	MPIO_Completed_request_create(&fd, len, error_code, request);
+    }
+#endif
 
     /* immediate completion */
     if (ret == 1) {

